google-cloud-pubsub/LOGGING.md

```diff
@@ -30,3 +33,87 @@ module GRPC
   extend MyLogger
 end
 ```
+
+## Adding gRPC interceptors
+
+[gRPC](https://github.com/grpc/grpc/tree/master/src/ruby) accepts [Ruby-language
+interceptors](https://github.com/grpc/proposal/blob/master/L11-ruby-interceptors.md) that allow you to insert your own
+custom logging into a client's RPC calls. (gRPC interceptors are also useful for auth, metrics, tracing and similar
+use cases.)
+
+This library performs RPCs using the following [gapic](https://github.com/googleapis/gapic-generator-ruby) clients from
+the underlying
+[google-cloud-pubsub-v1](https://github.com/googleapis/google-cloud-ruby/tree/main/google-cloud-pubsub-v1) library:
+
+* [`Google::Cloud::PubSub::V1::IAMPolicy::Client`](https://googleapis.dev/ruby/google-cloud-pubsub-v1/latest/Google/Cloud/PubSub/V1/IAMPolicy/Client.html)
+* [`Google::Cloud::PubSub::V1::Publisher::Client`](https://googleapis.dev/ruby/google-cloud-pubsub-v1/latest/Google/Cloud/PubSub/V1/Publisher/Client.html)
+* [`Google::Cloud::PubSub::V1::SchemaService::Client`](https://googleapis.dev/ruby/google-cloud-pubsub-v1/latest/Google/Cloud/PubSub/V1/SchemaService/Client.html)
+* [`Google::Cloud::PubSub::V1::Subscriber::Client`](https://googleapis.dev/ruby/google-cloud-pubsub-v1/latest/Google/Cloud/PubSub/V1/Subscriber/Client.html)
+
+To add a gRPC interceptor to one or more of these clients, first implement your logic as a subclass of
+[`GRPC::ClientInterceptor`](https://www.rubydoc.info/gems/grpc/GRPC/ClientInterceptor).
+
+```ruby
+require "grpc"
+require "logger"
+require "securerandom"
+
+class MyInterceptor < GRPC::ClientInterceptor
+  attr_reader :name
+
+  def initialize name
+    @name = name
+  end
+
+  def request_response(request:, call:, method:, metadata:)
+    logger.info "[#{name}] Sending unary request/response to #{method}"
+    metadata['request_id'] = generate_request_id
+    yield
+  end
+
+  def client_streamer(requests:, call:, method:, metadata:)
+    logger.info "[#{name}] Sending client streamer to #{method}"
+    metadata['request_id'] = generate_request_id
+    yield
+  end
+
+  def server_streamer(request:, call:, method:, metadata:)
+    logger.info "[#{name}] Sending server streamer to #{method}"
+    metadata['request_id'] = generate_request_id
+    yield
+  end
+
+  def bidi_streamer(requests:, call:, method:, metadata:)
+    logger.info "[#{name}] Sending bidi streamer to #{method}"
+    metadata['request_id'] = generate_request_id
+    yield
+  end
+
+  private
+
+  def logger
+    @logger ||= Logger.new(STDOUT)
+  end
+
+  def generate_request_id
+    SecureRandom.uuid
+  end
+end
+```
+
+Next, use the block yielded by a `Client.configure` method to add an instance of your class
+to the `interceptors` configuration of one or more of the generated clients.
```

Comment:
Add a note that the configurations must be set before the pubsub client is instantiated.

Maybe also add a few sentences noting which gapic clients need to be configured for which use cases. Otherwise, there are 4 gapic clients and users would have to dig into the veneer source to determine which ones are relevant to configure.

Reply:
> Maybe also add a few sentences noting which gapic clients need to be configured for which use cases.

I considered this, but it's tricky and time-consuming. I want to add this section for every gRPC-based handwritten client, so in the interest of time I was thinking that listing the gapic clients (with links) is an OK compromise. I just updated this sentence to refer to the list above.